clustering/marshalling/spi/src/main/java/org/wildfly/clustering/marshalling/spi/ByteBufferMarshalledValue.java

```diff
@@ -64,11 +65,18 @@ public synchronized boolean isEmpty() {
 
     public synchronized ByteBuffer getBuffer() throws IOException {
         ByteBuffer buffer = this.buffer;
-        if (buffer != null) return buffer;
-        if (this.object == null) return null;
-        ByteBuffer result = this.marshaller.write(this.object);
-        LOGGER.debugf("Marshalled size of %s(%s) = %d bytes", this.object.getClass().getCanonicalName(), this.object, result.limit() - result.arrayOffset());
-        return result;
+        if ((buffer == null) && (this.object != null)) {
+            // Since the wrapped object is likely mutable, we cannot cache the generated buffer
+            buffer = this.marshaller.write(this.object);
+            // N.B. Refrain from logging wrapped object
+            // If wrapped object contains an EJB proxy, toString() will trigger an EJB invocation!
+            LOGGER.debugf("Marshalled size of %s object = %d bytes", this.object.getClass().getCanonicalName(), buffer.limit() - buffer.arrayOffset());
+        }
+        return buffer;
+    }
+
+    public synchronized OptionalInt size() {
+        return (this.buffer != null) ? OptionalInt.of(this.buffer.limit() - this.buffer.arrayOffset()) : this.marshaller.size(this.object);
```

Comment:
A comment explaining this algorithm would be good. It's not clear why passing a buffer into the constructor (so this.buffer != null) means the state of that buffer means this.buffer.arrayOffset() is safe to call (i.e. it's a heap buffer) or that this.buffer.limit() - this.buffer.arrayOffset() is the correct value when size() is called.

Reply:
Valid point.  In practice, these are always array backed buffers - but we should always have hasArray() checks (and alternate handling) for completeness.

Reply:
Fixed.

Reply:
Oh - I see what you mean now.  Yes - I will add comments clarifying the lifecycle of this property, which is invalidated once the object is restored from the buffer.

Reply:
Done.